### PR TITLE
add feature flags for docs and types generator migration

### PR DESCRIPTION
**Pull request by phani-srikar**
*Description of changes:*
These changes add 2 Feature flags namely `codegen.useDocsGeneratorPlugin ` and `codegen.useTypesGeneratorPlugin ` which would help customers to switch back to the legacy versions of statements and types generator packages in the CLI repo. For newly created projects, the migrated packages will be used instead of the legacy ones in CLI repo.

The unit tests using the legacy packages are moved to `legacy-tests` folder which has the following advantages:
- Once the Feature flags are deprecated, we can simply delete the folder with minimal changes to actual tests.
- Testing any API changes/additions which are not supported by legacy packages can be done in a more isolated way. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-codegen/tests/commands/statements.test.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 const path = require('path');
-const { generate } = require('amplify-graphql-docs-generator');
+const { generate } = require('@aws-amplify/graphql-docs-generator');
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/statements.test.js:2]: Why is this line get replaced? The similar part in the type generation test but is not replaced.

File: packages/amplify-codegen/tests/callbacks/postPushCallback.test.js [hunk 2]
```diff
@@ -14,6 +14,22 @@ jest.mock('../../src/codegen-config');
...
 jest.mock('../../src/commands/types');
+// Mock the Feature flags for statements and types generation to use migrated packages
+jest.mock('amplify-cli-core', () => {
+  return {
+    FeatureFlags: {
+      getBoolean: jest.fn().mockImplementation((name, defaultValue) => {
+        if (name === 'codegen.useDocsGeneratorPlugin') {
+          return true;
+        }
+        if (name === 'codegen.usetypesGeneratorPlugin') {
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/callbacks/postPushCallback.test.js:25]: A typo here `useTypesGeneratorPlugin`

File: packages/amplify-codegen/tests/commands/add.test.js [hunk 3]
```diff
@@ -23,6 +23,21 @@ jest.mock('../../src/commands/types');
...
 jest.mock('../../src/utils');
+// Mock the Feature flags for statements and types generation to use migrated packages
+jest.mock('amplify-cli-core', () => {
+  return {
+    FeatureFlags: {
+      getBoolean: jest.fn().mockImplementation((name, defaultValue) => {
+        if (name === 'codegen.useDocsGeneratorPlugin') {
+          return true;
+        }
+        if (name === 'codegen.usetypesGeneratorPlugin') {
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/add.test.js:34]: The same typo here

File: packages/amplify-codegen/tests/commands/configure.test.js [hunk 4]
```diff
@@ -7,6 +7,22 @@ jest.mock('../../src/commands/add');
...
 
+// Mock the Feature flags for statements and types generation to use migrated packages
+jest.mock('amplify-cli-core', () => {
+  return {
+    FeatureFlags: {
+      getBoolean: jest.fn().mockImplementation((name, defaultValue) => {
+        if (name === 'codegen.useDocsGeneratorPlugin') {
+          return true;
+        }
+        if (name === 'codegen.usetypesGeneratorPlugin') {
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/configure.test.js:18]: Same typo

File: packages/amplify-codegen/tests/commands/generateStatementsAndTypes.test.js [hunk 5]
```diff
@@ -21,6 +21,22 @@ jest.mock('../../src/commands/types');
...
 jest.mock('../../src/utils');
+// Mock the Feature flags for statements and types generation to use migrated packages
+jest.mock('amplify-cli-core', () => {
+  return {
+    FeatureFlags: {
+      getBoolean: jest.fn().mockImplementation((name, defaultValue) => {
+        if (name === 'codegen.useDocsGeneratorPlugin') {
+          return true;
+        }
+        if (name === 'codegen.usetypesGeneratorPlugin') {
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/generateStatementsAndTypes.test.js:32]: Typo +1

File: packages/amplify-codegen/tests/commands/statements.test.js [hunk 6]
```diff
@@ -17,10 +17,26 @@ const MOCK_CONTEXT = {
...
 jest.mock('fs-extra');
+// Mock the Feature flags for statements and types generation to use migrated packages
+jest.mock('amplify-cli-core', () => {
+  return {
+    FeatureFlags: {
+      getBoolean: jest.fn().mockImplementation((name, defaultValue) => {
+        if (name === 'codegen.useDocsGeneratorPlugin') {
+          return true;
+        }
+        if (name === 'codegen.usetypesGeneratorPlugin') {
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/statements.test.js:32]: Typo +2

File: packages/amplify-codegen/tests/commands/types.test.js [hunk 7]
```diff
@@ -23,6 +23,22 @@ jest.mock('amplify-graphql-types-generator');
...
 jest.mock('fs-extra');
+// Mock the Feature flags for statements and types generation to use migrated packages
+jest.mock('amplify-cli-core', () => {
+  return {
+    FeatureFlags: {
+      getBoolean: jest.fn().mockImplementation((name, defaultValue) => {
+        if (name === 'codegen.useDocsGeneratorPlugin') {
+          return true;
+        }
+        if (name === 'codegen.usetypesGeneratorPlugin') {
```
Comment (AaronZyLee) [inline packages/amplify-codegen/tests/commands/types.test.js:34]: Typo +3

File: packages/amplify-codegen/tests/commands/statements.test.js (hunk 1 above)
Comment (AaronZyLee) [reply packages/amplify-codegen/tests/commands/statements.test.js:2]: I would suggest to have both package required and test it by switching the FF value to see if both `generate` gets called.

**Review by AaronZyLee — CHANGES_REQUESTED** [review]

File: packages/amplify-codegen/tests/commands/statements.test.js (hunk 1 above)
Comment (phani-srikar) [reply packages/amplify-codegen/tests/commands/statements.test.js:2]: I moved the legacy tests to a separate folder which can simply be deleted post deprecation of Feature flags. So these files now only test the API contracts between `amplify-codegen` and migrated packages.

**Review by AaronZyLee — APPROVED** [review]
LGTM 🌮

**Review by akshbhu — APPROVED** [review]
Discussed Offline , LGTM!

**Pull request merged**